crates/term/src/widget/mod.rs

```diff
@@ -0,0 +1,531 @@
+            CursorShape::HollowBlock => (egui::Color32::TRANSPARENT, config.theme.cursor_color),
+            CursorShape::Hidden => (egui::Color32::TRANSPARENT, egui::Color32::TRANSPARENT),
+        };
+
+        let blink = match config.cursor_blinking {
+            CursorBlinking::Always => true,
+            CursorBlinking::Never => false,
+            CursorBlinking::Terminal => cursor_style.blinking,
+        };
+
+        if blink {
+            let sin_component = self.stable_time / std::f32::consts::FRAC_PI_2 * 13.;
+            let alpha = (sin_component.sin() + 1.) / 2.;
+            inner_color = inner_color.gamma_multiply(alpha);
+        }
+
+        let cursor = galley.from_rcursor(egui::epaint::text::cursor::RCursor {
+            row: cursor_point.line.0 as usize,
+            column: cursor_point.column.0,
+        });
+
+        let mut cursor_pos = galley.pos_from_cursor(&cursor).min + response.rect.min.to_vec2();
+
+        let cursor_rect = match cursor_shape {
+            CursorShape::Block | CursorShape::HollowBlock | CursorShape::Hidden => {
+                egui::Rect::from_min_size(cursor_pos, egui::vec2(char_width, row_height))
+            }
+            CursorShape::Beam => egui::Rect::from_min_size(cursor_pos, egui::vec2(2.0, row_height)),
+            CursorShape::Underline => {
+                cursor_pos.y += row_height - 2.0;
+                egui::Rect::from_min_size(cursor_pos, egui::vec2(char_width, 2.0))
+            }
+        };
+
+        painter.rect(
+            cursor_rect,
+            egui::Rounding::ZERO,
+            inner_color,
+            egui::Stroke::new(1.0, outer_color),
+        );
+
+        // scrollbar
+        let min = response.rect.min + egui::vec2(galley.rect.width(), 0.0);
+        let size = egui::vec2(5., response.rect.height());
+        let sidebar_rect = egui::Rect::from_min_size(min, size);
+        painter.rect_filled(
+            sidebar_rect,
+            egui::Rounding::ZERO,
+            ui.visuals().extreme_bg_color,
+        );
+
+        // handle
+        let scrollbar_percent = screen_lines as f32 / total_lines as f32;
+        let scrollbar_height = scrollbar_percent * response.rect.height();
+
+        let scrollbar_offset_percent = display_offset as f32 / total_lines as f32;
+        let scrollbar_offset = scrollbar_offset_percent * response.rect.height();
+        let scrollbar_y = response.rect.max.y - scrollbar_offset - scrollbar_height;
+
+        let min = egui::pos2(min.x, scrollbar_y);
+        let size = egui::vec2(5., scrollbar_height);
+        let scrollbar_rect = egui::Rect::from_min_size(min, size);
+
+        painter.rect_filled(
+            scrollbar_rect,
+            egui::Rounding::same(5.),
+            ui.visuals().widgets.active.fg_stroke.color,
+        );
+
+        painter
+            .ctx()
+            .request_repaint_after(std::time::Duration::from_millis(16));
```

Comment:
It would drastically lower GPU usage to render the screen only when the contents of the terminal have updated rather than every frame. You can clone an `egui::Context` and call `.request_repaint()` from anywhere when the process sends new bytes.

To make this work for the cursor blinking, you can make the cursor change instantly in opacity every second or so rather than fading it and then change this `.request_repaint_after()` call to the cursor blink interval. This could be configurable.

Reply:
Caching layout jobs doesn't really fix the high GPU usage, and it also breaks scrolling

Reply:
How does it break scrolling? The ui only repaints every second or every time a terminal gets a `Wakeup` event as well, so I'm surprised that the gpu usage isn't any lower

Reply:
Looks like the ui is repainting pretty often, despite my changes, I'm looking into it

Reply:
Nevermind, the turns out the ui repaints ever frame when a toast is on screen